File: Baekjoon/Encryption/Hashing(15829).py
import sys
input = sys.stdin.readline

def powM(r: int, i: int, M: int):
  result = 1
  for _ in range(i):
    result = (result * r) % M
  return result

def main():
  r = 31
  M = 1234567891
  L = int(input())
  chars = input().strip()

  sum = 0
  for i, char in enumerate(chars):
    code = ord(char)-96
    sum = (sum + (code * powM(r, i, M)) % M) % M

  print(sum)

main()

File: Baekjoon/Encryption/Hashing(15829).js
/**
 * 
 * input string => alphabet lower case only
 * a=>1, b=>2, ..., z=>26
 * 
 * H = SUM / M
 * 해시충돌!
 * 
 * H = SUM(a*(r^i)) / M
 * 이때, r과 M은 서로소
 * 
 * r = 31
 * M = 1234567891
 * 
 */


const rl = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout,
});

const input = [];

rl.on('line', (line) => {
    input.push(line);
});

rl.on('close', () => {
    const r = 31;
    const M = 1234567891;
    const L = +input[0];
    
    let sum = 0;
    for (let i=0; i<L; i++) {
        const code = input[1][i].charCodeAt() - 96;
        sum = (sum + (code * pow(r, i, M)) % M) % M;
    }

    console.log(sum);
});

const pow = (r, i, M) => {
    let result = 1;
    for (let n=0; n<i; n++) {
        result = (result * r) % M
    }

    return result;
}